modules/renter/contractor/contracts.go

```diff
@@ -340,69 +340,163 @@ func (c *Contractor) threadedContractMaintenance() {
 	// hostdb.
 	c.managedMarkContractsUtility()
 
-	// Renew any contracts that need to be renewed.
-	c.mu.RLock()
-	var renewSet []types.FileContractID
-	// refundSet is used to mark contract that we want to 'refund': contracts we
-	// want to renew but leave the end heights unchanged.
-	refundSet := make(map[types.FileContractID]struct{})
-	for _, contract := range c.contracts {
-		if !contract.GoodForRenew {
-			continue
+	// Figure out which contracts need to be renewed, and while we have the
+	// lock, figure out the end height for the new contracts and also the amount
+	// to spend on each contract.
+	//
+	// refreshSet is used to mark contracts that need to be refreshed because
+	// they have run out of money. The refreshSet indicates how much currency
+	// was used in the previous contract which got exhausted, so that we know to
+	// put more money towards the refreshed contract.
+	//
+	// The actions inside this RLock are complex enough to merit wrapping them
+	// in a function where we can defer the unlock.
+	var endHeight types.BlockHeight
+	var fundsAvailable types.Currency
+	var renewSet []types.FileContractID // all contracts which need renewal, including refreshSet contracts
+	var renewAmounts []types.Currency   // the amount of funding to use when renewing
```

Comment:
Since each `id` in the `renewSet` should have an associated `amount` at the same index, it might be useful to add a sanity check after these slices are populated to verify that they have the same length.

Reply:
would `map[types.FileContractID]types.Currency work?`? Or perhaps
```go
var renewals []struct{
    id     types.FileContractID
    amount types.Currency
}